### PR TITLE
Allow `Json` type for `Query`, `Path` and `Header`.

### PR DESCRIPTION
**Pull request by luckydonald**
See #884 for details.

### CODE REVIEW

File: fastapi/dependencies/utils.py [hunk 1]
```diff
@@ -276,7 +277,7 @@ def get_dependant(
         param_field = get_param_field(
             param=param, default_field_info=params.Query, param_name=param_name
         )
-        if param_name in path_param_names:
+        if param_name in path_param_names and not issubclass(param.annotation, JsonWrapper):
```
Comment (dmontagu) [inline fastapi/dependencies/utils.py]: I defer to @tiangolo, but I personally don't think we should add special handling for `JsonWrapper` here or below.

I don't believe this is a performance sensitive part of the codebase, but adding special handling for this type:

1. Adds an on-going maintenance burden where we have to have tests for `JsonWrapper` and non-`JsonWrapper` code paths
2. Adds an on-going maintenance burden where we need to handle this edge case any time we may want to refactor this code in the future
3. Opens the door for requests for special handling of other types in the future

Moreover, because pydantic now runs its tests against FastAPI, it could potentially make it more difficult for pydantic to refactor the `Json` class in the future.

I could be convinced otherwise if this was a common request, but so far that doesn't seem to be the case.

There are potentially alternative approaches to make this work though; I'll comment on them elsewhere.

File: fastapi/dependencies/utils.py [hunk 2]
```diff
@@ -276,7 +277,7 @@ def get_dependant(
...
             assert is_scalar_field(
```
Comment (dmontagu) [inline fastapi/dependencies/utils.py]: I think it would be better if we could rework this function and/or code path (and the corresponding one below) in such a way that `Json[X]` fields would be handled properly.

In particular, it would be better if we could test a different (closely-related) condition that didn't require specifically checking for `Json[X]`-based field types. I don't know if such a condition exists or could be easily made to exist though.

-----

It may also just make sense to loosen the restriction on what is allowed as a path/query/header-parameter type; I defer to @tiangolo about that.

------

Another preferable approach, in my opinion, would be to just change pydantic in such a way that `Json[X]` types pass the check you want to perform. I haven't thought enough about the `Json` type to know whether there is a reasonable change that could be made here.

Assuming there *is* a reasonable change, depending on what is required, it may be a breaking change requiring waiting until pydantic v2. But if such a change is possible (whether reasonable or not), there *may* be a simple way to accomplish your goal by subclassing the `Json` class and changing an attribute (you could call the subclass `QueryJson` or similar).

Comment (luckydonald) [reply fastapi/dependencies/utils.py]: Is @tiangolo maintainer there as well which could make an change to pydantic for fastapi realistic?
In any case I'm not sure if Pydantic is the correct place needing a change.
After all Pydantic is handling the Json type very well, decoding json, and reading it into lists, dicts and if specified even into real models just fine.
Currently we are throwing away that capability by limiting it to just the Body as source type.

It's just FastAPI adding too restrictive code on top of pydantic, therefore I made an exception for the case we want to handle here, Json.

It might be possible to maybe loosen the existing code in a more general way, however I don't really understand the existing code _(a few comments here and there would be really nice!)_ to make any in depth changes.

**Comment by tiangolo** [comment]
Thanks @luckydonald for your contribution and interest. :cake: 

But I think this would be better handled in a dependency that parses the incoming string as @dmontagu suggested in the issue. That should solve it without needing extra logic in FastAPI for that specific use case. Also, the default encoding for query parameters is `form`, so this could end up misguiding users as it wouldn't do all the extra logic necessary to document it as a custom encoding in OpenAPI. As it is more of a corner case, I think the best approach is to use a dependency that does the JSON decoding from the string in the query.

Thanks @dmontagu for the review and detailed analysis! :rocket:
